File: backend/chat/events.go
package main

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/getsentry/sentry-go"
	"github.com/go-redis/redis/v7"
)

const notifyChannelName = "TSG_NOTIFY"

var userCountEvents = make(map[int]int)

// Event represents an event which is meant to be sent to an user.
type Event struct {
	UserID int    `json:"user_id"`
	Typ    string `json:"typ"`
	Msg    string `json:"msg"`
}

func (e *Event) String() string {
	return fmt.Sprintf("User-id: %d, Typ: %s", e.UserID, e.Typ)
}

// Watches in redis the notfiy list.
// The django backend stores event which are meant to be sent
// a specific user.
//
// If an event gets stored in the list in redis,
// watchRedisNotifiyList() consumes the event.
//
// Afterwards, the event gets sent to the user, if
// the user is currently online and has an active websocket connection.
func watchRedisNotifiyList() {

	log.Info("Connecting with Redis...")
	client, err := connectRedis()

	if err != nil {
		log.Fatalf("Failed to connect with redis, err: %s\n", err)
		sentry.CaptureException(err)
	}

	log.Infoln("Successfully connected with Redis, now listening for tasks!")

	// Start goroutine which resets the maximum of events per user
	go func() {
		log.Infoln("Started go routine for cleaning the limits for each user!")
		for {
			userCountEvents = make(map[int]int)
			time.Sleep(time.Millisecond * 500)
		}
	}()

	for {

		result, err := client.BLPop(0, notifyChannelName).Result()

		if err != nil {
			log.Fatalf("Could not read from redis list, err: %s\n", err)
			sentry.CaptureException(err)
		}

		var event Event
		err = json.Unmarshal([]byte(result[1]), &event)
		if err != nil {
			log.Warnf("Could not unmarshal json: %s\n", err)
			sentry.CaptureException(err)
			continue
		}

		// We could also think about accumlating events per user id in a map.
		// Then start for each id in the map a goroutine which sends the events in a given rate.
		if !limitEventsReached(event.UserID) {
			log.Infof("Got Event: %s. Now sending event to user\n", event.String())
			sendEvent(event)
		} else {
			log.Infof("User with id %d already got the maximum of events. Not sending event!\n", event.UserID)
		}
	}

}

// limitEventsReached returns true if a user has reached his maximum number of notifications
// and thus he/she should not reach another notifciation.
//
// The reason for that is, that we dont want to flood the user with notifications.
func limitEventsReached(userID int) bool {

	_, ok := userCountEvents[userID]
	if !ok {
		userCountEvents[userID] = 1
		return false
	}

	userCountEvents[userID]++

	return userCountEvents[userID] > 5

}

func connectRedis() (*redis.Client, error) {
	client := redis.NewClient(&redis.Options{

		// TODO: Read from env
		Addr:     getEnvOrDefaultValue("REDIS_ADDR"),
		Password: getEnvOrDefaultValue("REDIS_PASSWORD"), // no password set
		DB:       0,                                      // use default DB
	})
	_, err := client.Ping().Result()
	return client, err
}

// sendEvent sends an event to the user specified in the event
func sendEvent(event Event) {
	ws, ok := connections[event.UserID]
	if !ok {
		log.Infof("User with id %d not connected\n", event.UserID)
		return
	}

	sendToWs(event, EventMsg, ws)

}

File: backend/notify/events.py
import redis
import logging
import json

from tsg import settings

NOTIFY_CHANNEL_NAME = "TSG_NOTIFY"

logger = logging.getLogger(__name__)

r = redis.Redis.from_url(settings.REDIS_URL)


class Event:
    """
    An event which is meant to be sent to a user via websocket
    """

    def __init__(self, user_id: int, typ: str, msg: str):
        self.user_id = user_id
        self.typ = typ
        self.msg = msg

    def to_json(self) -> str:
        return json.dumps(self, default=lambda o: o.__dict__)


def store_event(event: Event) -> None:
    """
    Stores an event in the redis database.

    The stored event then gets consumed by the go worker who proceeds to send
    the event to the user via a websocket connection if the user is connected.
    """

    if isinstance(event, Event) is False:
        logger.error(f"Can only save objects of typ {Event.__class__}. Object was of class {event.__class__}")
        return
    try:
        r.ping()
    except redis.exceptions.ConnectionError:
        logger.error("Could not connect with redis!")
        return

    if not r.lpush(NOTIFY_CHANNEL_NAME, event.to_json()):
        logger.error(f"Failed to push {event} to {NOTIFY_CHANNEL_NAME} queue!")
